File: packages/capacitor-proximity/android/src/main/java/com/praeventio/proximity/CapacitorProximityPlugin.java
package com.praeventio.proximity;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorProximity")
public class CapacitorProximityPlugin extends Plugin {
    private CapacitorProximity implementation;

    @Override
    public void load() {
        super.load();
        implementation = new CapacitorProximity(
            getActivity(),
            reading -> notifyListeners("proximityChanged", reading)
        );
    }

    @Override
    protected void handleOnDestroy() {
        if (implementation != null) {
            implementation.stopMonitoring();
        }
        super.handleOnDestroy();
    }

    @PluginMethod
    public void enable(PluginCall call) {
        bridge.executeOnMainThread(() -> implementation.enable(call));
    }

    @PluginMethod
    public void disable(PluginCall call) {
        bridge.executeOnMainThread(() -> implementation.disable(call));
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        call.resolve(implementation.getStatus());
    }

    @PluginMethod
    public void getCurrent(PluginCall call) {
        JSObject current = implementation.getCurrent();
        if (current == null) {
            call.reject("No proximity reading is available yet.");
            return;
        }
        call.resolve(current);
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject result = new JSObject();
        result.put("version", implementation.getPluginVersion());
        call.resolve(result);
    }
}

File: packages/capacitor-proximity/ios/Sources/PraeventioCapacitorProximity/CapacitorProximityPlugin.swift
import Capacitor
import Foundation

@objc(CapacitorProximityPlugin)
public class CapacitorProximityPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorProximityPlugin"
    public let jsName = "CapacitorProximity"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "enable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = CapacitorProximity()

    deinit {
        implementation.disable()
    }

    @objc func enable(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let enabled = self.implementation.enable { [weak self] isNear, timestamp in
                self?.notifyListeners("proximityChanged", data: [
                    "state": isNear ? "near" : "far",
                    "timestamp": timestamp
                ])
            }
            enabled ? call.resolve() : call.reject("Proximity sensor not available on this device.")
        }
    }

    @objc func disable(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.implementation.disable()
            call.resolve()
        }
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "available": self.implementation.isAvailable(),
                "enabled": self.implementation.isEnabled(),
                "platform": "ios"
            ])
        }
    }

    @objc func getCurrent(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let state = self.implementation.currentState() else {
                call.reject("No proximity reading is available while monitoring is disabled.")
                return
            }
            call.resolve([
                "state": state.boolValue ? "near" : "far",
                "timestamp": Date().timeIntervalSince1970 * 1_000
            ])
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": implementation.getPluginVersion()])
    }
}
